x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlMetrics.java

```diff
@@ -0,0 +1,488 @@
+            "Sum of failed trained model allocations across all deployments cluster-wide.",
+            "allocations",
+            () -> new LongWithAttributes(trainedModelAllocationCounts.trainedModelsFailedAllocations, isMasterMap)
+        );
+    }
+
+    /**
+     * Metric values are recalculated in response to cluster state changes and then cached.
+     * This means that the telemetry provider can poll the metrics registry as often as it
+     * likes without causing extra work in recalculating the metric values.
+     */
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        isMasterMap = event.localNodeMaster() ? MASTER_TRUE_MAP : MASTER_FALSE_MAP;
+
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            // Wait until the gateway has recovered from disk.
+            return;
+        }
+
+        boolean mustRecalculateFreeMem = false;
+
+        final ClusterState currentState = event.state();
+        final ClusterState previousState = event.previousState();
+
+        if (firstTime || event.metadataChanged()) {
+            final PersistentTasksCustomMetadata tasks = currentState.getMetadata().custom(PersistentTasksCustomMetadata.TYPE);
+            final PersistentTasksCustomMetadata oldTasks = firstTime
+                ? null
+                : previousState.getMetadata().custom(PersistentTasksCustomMetadata.TYPE);
+            if (tasks != null && tasks.equals(oldTasks) == false) {
+                if (hasMasterRole) {
+                    mlTaskStatusCounts = findTaskStatuses(tasks);
+                }
+                if (hasMlRole) {
+                    nativeMemAdUsage = findAdMemoryUsage(autodetectProcessManager);
+                    nativeMemDfaUsage = findDfaMemoryUsage(dataFrameAnalyticsManager, tasks);
+                    mustRecalculateFreeMem = true;
+                }
+            }
+        }
+
+        final TrainedModelAssignmentMetadata currentMetadata = TrainedModelAssignmentMetadata.fromState(currentState);
+        final TrainedModelAssignmentMetadata previousMetadata = firstTime ? null : TrainedModelAssignmentMetadata.fromState(previousState);
+        if (currentMetadata != null && currentMetadata.equals(previousMetadata) == false) {
+            if (hasMasterRole) {
+                trainedModelAllocationCounts = findTrainedModelAllocationCounts(currentMetadata);
+            }
+            if (hasMlRole) {
+                nativeMemTrainedModelUsage = findTrainedModelMemoryUsage(currentMetadata, currentState.nodes().getLocalNode().getId());
+                mustRecalculateFreeMem = true;
+            }
+        }
+
+        if (firstTime) {
+            firstTime = false;
+            nativeMemLimit = findNativeMemoryLimit(currentState.nodes().getLocalNode(), clusterService.getClusterSettings());
+            mustRecalculateFreeMem = true;
+            clusterService.getClusterSettings()
```

Comment:
The reason they aren't in the constructor is that the `memoryLimitClusterSettingUpdated()` method triggers recalculation of `limit` and `free` _assuming_ that `used` is storing a relatively up-to-date value. If we calculate `free` too early then it will be set equal to `limit`, assuming `used` is zero.

I thought it would be better that immediately after node startup when the stats are all inaccurate that they're all spuriously zero rather than some zero and some not.

But I can see that this isn't obvious, so I'll add a code comment to explain.